File: VB/GridControlSaver/GCDataSaver.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.IO
Imports System.Data
Imports DevExpress.XtraGrid
Imports System.Collections
Imports System.Xml.Serialization
Imports System.Reflection
Imports System.Xml




Namespace GridControlSaver
	Public Class GCDataSaver
		Private gridContrl As GridControl


		Public Sub New(ByVal gridContrl As GridControl)
			Me.gridContrl = gridContrl
		End Sub



		Public Sub SaveToXml(ByVal fileName As String)
			If gridContrl Is Nothing Then
				Return
			End If

			Dim f_stream As New FileStream(fileName, FileMode.Create)
			Try
				Dim str As String = "<?xml version=""1.0""?>" & Constants.vbCrLf & "<container>" & Constants.vbCrLf
				Dim buffer() As Byte = Encoding.ASCII.GetBytes(str)
				f_stream.Write(buffer, 0, buffer.Length)
				gridContrl.DefaultView.SaveLayoutToStream(f_stream)

				SaveDataSource(f_stream)

				str = "</container>"
				buffer = Encoding.ASCII.GetBytes(str)
				f_stream.Write(buffer, 0, buffer.Length)
			Finally
				f_stream.Close()
			End Try
		End Sub




		Private Sub SaveDataSource(ByVal f_stream As FileStream)
			If gridContrl Is Nothing OrElse gridContrl.DataSource Is Nothing Then
				Return
			End If

			Dim ms As New MemoryStream()

			Try
				Dim xml_doc As New XmlDocument()

				Dim xml_serial As New XmlSerializer(gridContrl.DataSource.GetType())

				xml_serial.Serialize(ms, gridContrl.DataSource)
				ms.Position = 0
				xml_doc.Load(ms)

				Dim buffer() As Byte = Encoding.ASCII.GetBytes(xml_doc.DocumentElement.OuterXml)
				f_stream.Write(buffer, 0, buffer.Length)
			Catch
			Finally
				ms.Close()
			End Try
		End Sub



		Public Sub RestoreFromXml(ByVal fileName As String, ByVal dataSourceType As Type)
			If gridContrl Is Nothing Then
				Return
			End If

			Dim xml_doc As New XmlDocument()
			Dim m_stream As New MemoryStream()
			Try
				xml_doc.Load(fileName)
				Dim node_list As XmlNodeList = xml_doc.DocumentElement.ChildNodes
				Dim buffer() As Byte
				For i As Integer = 0 To node_list.Count - 1
					buffer = Encoding.ASCII.GetBytes(node_list(i).OuterXml)

					m_stream.Write(buffer, 0, buffer.Length)

					If i = 0 Then
						m_stream.Position = 0
						gridContrl.DefaultView.RestoreLayoutFromStream(m_stream)
						m_stream.SetLength(0)
					End If
				Next i

				If m_stream.Length > 0 Then
					m_stream.Position = 0
					Dim xml_serial As New XmlSerializer(dataSourceType)
					gridContrl.DataSource = xml_serial.Deserialize(m_stream)
				End If
			Catch
			Finally
				m_stream.Close()
			End Try
		End Sub
	End Class
End Namespace

File: CS/GridControlSaver/GCDataSaver.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Data;
using DevExpress.XtraGrid;
using System.Collections;
using System.Xml.Serialization;
using System.Reflection;
using System.Xml;




namespace GridControlSaver
{
    public class GCDataSaver
    {
        private GridControl gridContrl;


        public GCDataSaver(GridControl gridContrl)
        {
            this.gridContrl = gridContrl;
        }



        public void SaveToXml(string fileName)
        {
            if (gridContrl == null) return;

            FileStream f_stream = new FileStream(fileName, FileMode.Create);
            try
            {
                string str = "<?xml version=\"1.0\"?>\r\n<container>\r\n";
                byte[] buffer = Encoding.ASCII.GetBytes(str);
                f_stream.Write(buffer, 0, buffer.Length);
                gridContrl.DefaultView.SaveLayoutToStream(f_stream);

                SaveDataSource(f_stream);

                str = "</container>";
                buffer = Encoding.ASCII.GetBytes(str);
                f_stream.Write(buffer, 0, buffer.Length);
            }
            finally
            {
                f_stream.Close();
            }
        }




        private void SaveDataSource(FileStream f_stream)
        {
            if (gridContrl == null || gridContrl.DataSource == null) return;

            MemoryStream ms = new MemoryStream();
            
            try
            {
                XmlDocument xml_doc = new XmlDocument();

                XmlSerializer xml_serial = new XmlSerializer(gridContrl.DataSource.GetType());

                xml_serial.Serialize(ms, gridContrl.DataSource);
                ms.Position = 0;
                xml_doc.Load(ms);
            
                byte[] buffer = Encoding.ASCII.GetBytes(xml_doc.DocumentElement.OuterXml);
                f_stream.Write(buffer, 0, buffer.Length);
            }
            catch
            {
            }
            finally
            {
                ms.Close();
            }
        }



        public void RestoreFromXml(string fileName, Type dataSourceType)
        {
            if (gridContrl == null) return;

            XmlDocument xml_doc = new XmlDocument();
            MemoryStream m_stream = new MemoryStream();
            try
            {
                xml_doc.Load(fileName);
                XmlNodeList node_list = xml_doc.DocumentElement.ChildNodes;
                byte[] buffer;
                for (int i = 0; i < node_list.Count; i++)
                {
                    buffer = Encoding.ASCII.GetBytes(node_list[i].OuterXml);

                    m_stream.Write(buffer, 0, buffer.Length);

                    if (i == 0)
                    {
                        m_stream.Position = 0;
                        gridContrl.DefaultView.RestoreLayoutFromStream(m_stream);
                        m_stream.SetLength(0);
                    }
                }

                if (m_stream.Length > 0)
                {
                    m_stream.Position = 0;
                    XmlSerializer xml_serial = new XmlSerializer(dataSourceType);
                    gridContrl.DataSource = xml_serial.Deserialize(m_stream);
                }
            }
            catch
            {
            }
            finally
            {
                m_stream.Close();
            }
        }
    }
}
